### PR TITLE
FIX: remove racing requests for admin users

### PR DESCRIPTION
**Pull request by markvanlan**

### CODE REVIEW

**Comment by ZogStriP** [comment]
Shouldn't we solve the underlying problem instead? We do not seem to be cancelling the previous requests whenever the user types. All the previous requests are in flight and they all end up finishing and the line `this._results = this._results.concat(result);` ends up creating a lot of duplicates.

**Comment by markvanlan** [comment]
@ZogStriP You can't cancel requests though; you can ignore the response which is what this adds. In my opinion the underlying problem is that we are evaluating logic on the response of old requests. This returns early if we detect that the request is old

**Comment by ZogStriP** [comment]
> I do not agree that this solution doesn't fix the underlying issue

By the "underlying issue" I meant it as this problem most likely exist elsewhere in the application. You are only solving it for this specific request but maybe we should solve it deeper in the layers so we don't get this bug elsewhere?

**Comment by markvanlan** [comment]
That would probably mean adding some new API, or modifying our ajax lib which is just out of scope of this TODO... Shall we spin up a dev topic for that? I do think that would be really nice to add probably to our `ajax` implementation.

**Comment by pmusaraj** [comment]
Small bump here @markvanlan. We use Ember's runloop `cancel` in a few places to cancel debounced ajax calls (see `search-menu` or `text-field`). This feels very similar to how `resetFilters` is called here, it does look to me like that should work.

**Comment by markvanlan** [comment]
@CvX ^ You remember this conversation in chat?

**Comment by SamSaffron** [comment]
OK ... looks like we are not happy with this specific fix @pmusaraj / @markvanlan / @CvX / @ZogStriP   I am closing this so we can try something else? 

Also a bit odd that we have "refreshing" thing, should it not be banning these extra calls?

**Comment by markvanlan** [comment]
@SamSaffron No, it's fine if it's refreshing before responses come back; the issue is simply canceling the old requests to make sure that we always have the last request... requested, evaluated. If the TODO is to fix the issue here I still stand by the fix as @CvX corrected me when I tried to use `cancel` to cancel an ajax request. If we want a general solution for this we will need some wrapper around ajax, and that hasn't been made clear to me still.

**Comment by CvX** [comment]
Canceling `debounce(resetFilters)` would not work. Depending on where you'd place that cancel, it would either not do anything (because the request could be already in-flight) or prevent the new search from triggering when an old request finishes… So that idea is a no-go.

I've looked at how we actually handle similar issues :eyes: and it looks like we don't really do much. All we do is relying on canceling debounced handlers, but that won't do much if there's any async/network requests…

The reason this this particular case was noticed is because the requests here can take exceptionally long time to finish (in most pathological situations even 2s+ to return just a single record!), and that we append results instead of replacing them which makes the issue easier to spot.

Regis suggested `AbortController` which we will most likely use - but in the future. For now, we're still relying on jQuery's ajax, and abort controller works only with `fetch` API.

So… **I think this PR is actually okay for now.** 

---

We can loop around to the wider discussion about the issue of cancelling previous request after we phase out ajax from the app (which is on my todo list, but as always, it's a rabbit hole)

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 1]
```diff
@@ -17,13 +17,15 @@ export default Controller.extend(CanCheckEmails, {
   listFilter: null,
   selectAll: false,
   searchHint: i18n("search_hint"),
+  _searchIndex: null,
```
Comment (CvX) [inline app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:20]:
```suggestion
  _searchIndex: 0,
```

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 2]
```diff
@@ -17,13 +17,15 @@ export default Controller.extend(CanCheckEmails, {
...
 
   init() {
     this._super(...arguments);
 
     this._page = 1;
     this._results = [];
     this._canLoadMore = true;
+    this._searchIndex = 0;
```
Comment (CvX) [inline app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:28]:
```suggestion
```

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 3]
```diff
@@ -58,14 +62,22 @@ export default Controller.extend(CanCheckEmails, {
       page: this._page,
     })
       .then((result) => {
+        if (searchIndex !== this._searchIndex) {
```
Comment (CvX) [inline app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:65]:
```suggestion
        if (searchIndex !== this._searchIndex || this.isDestroyed || this.isDestroying) {
```

File: app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js [hunk 4]
```diff
@@ -58,14 +62,22 @@ export default Controller.extend(CanCheckEmails, {
...
         }
 
         this._results = this._results.concat(result);
         this.set("model", this._results);
       })
-      .finally(() => this.set("refreshing", false));
+      .finally(() => {
+        if (searchIndex === this._searchIndex) {
+          this.set("refreshing", false);
+        }
```
Comment (CvX) [inline app/assets/javascripts/admin/addon/controllers/admin-users-list-show.js:79]:
```suggestion
        if (searchIndex !== this._searchIndex || this.isDestroyed || this.isDestroying) {
          return;
        }

        this.set("refreshing", false);
```

**Review by CvX — APPROVED** [review]

**Review by CvX — APPROVED** [review]

**Pull request merged**